Removed GlobalParams applying to suppressed parameters

socratesgorilla opened the pull request:
Closes #8761

Commit:
Removed GlobalParams applying to suppressed parameters #8761

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/1223912/) on 87580a1 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/22592/site)

This comment will be updated on new commits.

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/1223925/) on 87580a1 wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>c8ebf4</th>
      <th colspan=4>#22592 87580a</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>84.36%</td>
      <td>84.36%</td>
      <td>+0.00%</td>
      <td>100.00%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>80674</td>
      <td>80678</td>
      <td>+4</td>
      <td>1</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>14956</td>
      <td>14952</td>
      <td>-4</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/22592/coverage_diff/framework)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/22592/coverage/framework)


## Modules coverage

Coverage did not change

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/22592/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/22592/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/22592/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/22592/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/22592/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/22592/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/22592/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/22592/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/22592/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/22592/coverage/geochemistry)
- [`heat_conduction`](https://mooseframework.inl.gov/docs/PRs/22592/coverage/heat_conduction)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/22592/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/22592/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/22592/coverage/navier_stokes)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/22592/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/22592/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/22592/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/22592/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/22592/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/22592/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/22592/coverage/richards)
- [`scalar_transport`](https://mooseframework.inl.gov/docs/PRs/22592/coverage/scalar_transport)
- [`solid_properties`](https://mooseframework.inl.gov/docs/PRs/22592/coverage/solid_properties)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/22592/coverage/stochastic_tools)
- [`tensor_mechanics`](https://mooseframework.inl.gov/docs/PRs/22592/coverage/tensor_mechanics)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/22592/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/22592/coverage/xfem)

</details>



This comment will be updated on new commits.

GiudGiud commented on test/tests/globalparams/global_param/tests:
```diff
@@ -1,12 +1,20 @@
+    input = 'suppresscheck.i'
+    expect_err = "The parameter 'Adaptivity/Markers/boundary/use_displaced_mesh' is a private parameter and should not be used in an input file."
+
+    requirement = 'The system shall not apply global parameter options onto suppressed parameters.'
```
this is not the test you want for this. 

This is a good test for "The system shall error if a private parameter is set in the input file"

A good test for that is a RunApp test that does not error

GiudGiud replied:
you can keep this test, it does not already exist

GiudGiud commented on test/tests/globalparams/global_param/suppresscheck.i:
```diff
@@ -0,0 +1,67 @@
+      distance = 0.35
+      mark = refine
+
+      #suppressed params should still error properly when defined locally in their objects, so we set it true here
+      use_displaced_mesh = true
```
pass it from the command line so you can reuse the input file for both tests

GiudGiud commented on framework/src/parser/Parser.C:
```diff
@@ -1255,7 +1255,7 @@ Parser::extractParams(const std::string & prefix, InputParameters & p)
     else
     {
-      if (p.isPrivate(it.first))
+      if (p.isPrivate(it.first) and !in_global)
```
&& ??

socratesgorilla replied:
and is a valid cpp keyword

GiudGiud replied:
we dont use it though. Use &&, it matches the local style

YaqiWang commented:
I think the goal in @dschwen 's issue is to stop applying global parameters to suppressed parameters but not just issue an error unless I misunderstood.

GiudGiud commented:
If these changes apply the global parameter to the suppressed parameter this is problematic. 
I think we ll want to check that with a test as well.

Commit:
Modified global param tests, added new tests

GiudGiud commented on test/tests/globalparams/global_param/tests:
```diff
@@ -1,12 +1,28 @@
+    cli_args = "Adaptivity/Markers/boundary/use_displaced_mesh=true"
+    expect_err = "The parameter 'Adaptivity/Markers/boundary/use_displaced_mesh' is a private parameter and should not be used in an input file."
+    mesh_mode = REPLICATED
+    requirement = 'The system shall error on locally suppressed parameters even when a global parameter is being ignored.'
```
```suggestion
    requirement = 'The system shall error if a private / suppressed parameter is set in the input file.'
```
the global parameter is irrelevant here, local parameters always take precedence

GiudGiud commented on framework/src/parser/Parser.C:
```diff
@@ -1255,7 +1255,7 @@ Parser::extractParams(const std::string & prefix, InputParameters & p)
     else
     {
-      if (p.isPrivate(it.first))
+      if (p.isPrivate(it.first) && !in_global)
```
if it s private and in global, should we return here to avoid setting it ?

socratesgorilla replied:
Been too long that I don't have a strong opinion. I'll default to your judgement. Other issues taken care of.

GiudGiud replied:
> If these changes apply the global parameter to the suppressed parameter this is problematic.
> I think we ll want to check that with a test as well.
Basically if this is happening, which we dont want, then I think you need to return here to make it stop from happening

GiudGiud replied:
this being "the global parameter ends up being applied to the suppressed parameter"

GiudGiud replied:
Fixed now. It was indeed being applied

GiudGiud commented on test/tests/globalparams/global_param/tests:
```diff
@@ -1,12 +1,28 @@
+  
+  [test_suppress_ignore]
+    type = 'RunException'
+    input = 'suppresscheck.i'
```
rename this file to suppress_check.i 
it s the convention for all test inputs names

GiudGiud committed:
Add an actual check that the suppressed parameter is not being set by a global param
refs #8761

GiudGiud committed:
Do NOT set private parameters using global parameters, refs #8761

GiudGiud committed:
Private parameters also present in global params should not be considered set by the user

GiudGiud commented:
@lindsayad for the independent review please

lindsayad commented on test/include/userobjects/TestGlobalParamSuppression.h:
```diff
@@ -0,0 +1,24 @@
+
+  TestGlobalParamSuppression(const InputParameters & parameters);
+
+  virtual void initialize(){};
+  virtual void execute(){};
+  virtual void finalize(){};
```
```suggestion
  virtual void initialize() override {}
  virtual void execute() override {}
  virtual void finalize() override {}
```

lindsayad commented on test/tests/globalparams/global_param/suppress_check.i:
```diff
@@ -0,0 +1,58 @@
+
+[Executioner]
+  type = Steady
+
+  solve_type = 'PJFNK'
+
```
```suggestion
  solve_type = 'PJFNK'
```

lindsayad commented on test/tests/globalparams/global_param/tests:
```diff
@@ -1,12 +1,28 @@
+    mesh_mode = REPLICATED
+    requirement = 'The system shall not apply global parameter options onto suppressed parameters.'
+  []
+
```
```suggestion
```

GiudGiud committed:
Fix test failure in TM module : suppressed parameter is being checked for being set by user
which it is not
Same in framework
Setting a private param programmatically is not setting by user

Update test/include/userobjects/TestGlobalParamSuppression.h

Co-authored-by: Alex Lindsay <alexlindsay239@gmail.com>

Update test/tests/globalparams/global_param/suppress_check.i

Co-authored-by: Alex Lindsay <alexlindsay239@gmail.com>

Update test/tests/globalparams/global_param/tests

Co-authored-by: Alex Lindsay <alexlindsay239@gmail.com>

GiudGiud commented:
Failing test, some stuff going down with parameters in modules. I ll find fixes though

GiudGiud commented on framework/src/actions/CreateProblemDefaultAction.C:
```diff
@@ -76,7 +76,7 @@ CreateProblemDefaultAction::act()
       params.set<MooseMesh *>("mesh") = _mesh.get();
       params.set<bool>("use_nonlinear") = _app.useNonlinear();
-      if (_pars.isParamSetByUser("_solve"))
+      if (_pars.isParamValid("_solve"))
```
this used to say true for setting it in a unit test. 
But it s a suppressed param, so it was not really set by a user, it was set programmatically "as if a user" but through an API a user does not have access to. 

anyway this works now

GiudGiud commented:
Excited for this 6 year old issue to be solved haha

GiudGiud commented:
@lindsayad
